pdal/pio.py

```diff
@@ -0,0 +1,108 @@
+import types
+import json
+import subprocess
+from functools import partial
+from collections import defaultdict
+from itertools import chain
+import copy
+import warnings
+
+import pdal
+
+try:
+    PDAL_DRIVERS_JSON = subprocess.run(["pdal", "--drivers", "--showjson"], capture_output=True).stdout
+    PDAL_DRIVERS = json.loads(PDAL_DRIVERS_JSON)
+    _PDAL_VALIDATE = True
+except:
+    PDAL_DRIVERS = []
+    _PDAL_VALIDATE = False
+
+DEFAULT_STAGE_PARAMS = defaultdict(dict)
+DEFAULT_STAGE_PARAMS.update({
+# TODO: add stage specific default configurations
+})
+
+
+class StageSpec(object):
+    def __init__(self, prefix, **kwargs):
+        self.prefix = prefix
+        self.key = ".".join([self.prefix, kwargs.get("type", "")])
+        self.spec = DEFAULT_STAGE_PARAMS[self.key].copy()
+        self.spec.update(kwargs)
+        self.spec["type"] = self.key
+        # NOTE: special case to support reading files without passing an explicit reader
+        if (self.prefix == "readers") and kwargs.get("type") == "auto":
+            del self.spec["type"]
+
+    @property
+    def pipeline(self):
+        output = PipelineSpec()
+        output.add_stage(self)
+        return output
+
+    def __getattr__(self, name):
+        if _PDAL_VALIDATE and (name not in dir(self)):
+            raise AttributeError(f"'{self.prefix}.{name}' is an invalid or unsupported PDAL stage")
+        return partial(self.__class__, self.prefix, type=name)
+
+    def __str__(self):
+        return json.dumps(self.spec, indent=4)
+
+    def __add__(self, other):
+        return self.pipeline + other
+
+    def __dir__(self):
+        extra_keys = [e["name"][len(self.key):] for e in PDAL_DRIVERS if e["name"].startswith(self.key)] + ["auto"]
+        return super().__dir__() + [e for e in extra_keys if len(e) > 0]
+
+    def execute(self):
+        return self.pipeline.execute()
+
+
+readers = StageSpec("readers")
```

Comment:
Think these are now dead.

Reply:
I think I have to keep readers/writers/filters because the dot syntax within stage names conflicts with python module/class/attribute conventions.  I'm not sure if there's a more magical way getting different behavior off the top of my head but `pio.readers.ply` will lookup the attribute `readers` on `pio` and then the attribute `ply` on that rather than passing `readers.ply` to something directly.

Reply:
Yes, sorry.  I missed how these were being used.